waku/v2/peermanager/rtt_cache.go

```diff
@@ -0,0 +1,256 @@
+package peermanager
+
+import (
+	"container/heap"
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/discovery/backoff"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	"github.com/waku-org/go-waku/logging"
+	"github.com/waku-org/go-waku/waku/v2/timesource"
+	"go.uber.org/zap"
+)
+
+type PeerRTT struct {
+	pingQueueIndex int
+
+	PeerID  peer.ID
+	PingRTT time.Duration
+
+	bkf              backoff.BackoffStrategy
+	nextVerification time.Time
+	timesource       timesource.Timesource
+}
+
+func (pr *PeerRTT) IsStale() bool {
+	return pr.nextVerification.Before(pr.timesource.Now())
+}
+
+type RTTCache struct {
+	sync.RWMutex
+
+	host host.Host
+
+	peers     map[peer.ID]*PeerRTT
+	pingQueue PingQueue
+
+	verificationInterval time.Duration
+
+	timesource timesource.Timesource
+	logger     *zap.Logger
+}
+
+func NewRTTCache(timesource timesource.Timesource, logger *zap.Logger) *RTTCache {
+	return &RTTCache{
+		timesource: timesource,
+		logger:     logger.Named("ping-manager"),
+		peers:      make(map[peer.ID]*PeerRTT),
+	}
+}
+
+func (r *RTTCache) SetHost(h host.Host) {
+	r.host = h
+}
+
+func (r *RTTCache) Start(ctx context.Context) {
+	go r.start(ctx, 10*time.Second)
+}
+
+func (r *RTTCache) pingPeers(ctx context.Context) {
+	r.RLock()
+	var toPing peer.IDSlice
+	if len(r.pingQueue) == 0 {
+		toPing = r.host.Peerstore().Peers()
+	} else {
+		for _, p := range r.pingQueue {
+			if p.IsStale() {
+				toPing = append(toPing, p.PeerID)
```

Comment:
We may also have to handle removal of peers that are no longer in peerStore or peers that were previously connected, but not disconnected.

Reply:
I ended up checking if the peer to ping has info in the peerStore before adding it to `toPing`